File: 1333-sort-the-jumbled-numbers/sort-the-jumbled-numbers.py
class Solution:
    def sortJumbled(self, mapping: List[int], nums: List[int]) -> List[int]:
        # Step 1: Create a list to store original nums and their mapped values
        mapped_list = []
        for i, num in enumerate(nums):
            s = str(num)
            n = ''.join(str(mapping[int(ch)]) for ch in s)
            mapped_list.append((num, int(n), i))

        # Step 2: Sort the list based on the mapped values and original indices for stability
        mapped_list.sort(key=lambda x: (x[1], x[2]))

        # Step 3: Create a result list and fill it with the sorted original nums
        sorted_nums = [t[0] for t in mapped_list]

        return sorted_nums

# Example usage
sol = Solution()
mapping = [2, 1, 4, 8, 6, 3, 0, 9, 7, 5]
nums = [990, 332, 981, 330]
sorted_nums = sol.sortJumbled(mapping, nums)
print(sorted_nums)  # Output should be sorted based on mapped values, preserving order for duplicates

File: 1333-sort-the-jumbled-numbers/sort-the-jumbled-numbers.cpp
class Solution {
public:
    vector<int> sortJumbled(vector<int>& mapping, vector<int>& nums) {
        // Step 1: Create a vector to store original nums and their mapped values
        vector<tuple<int, int, int>> mappedList;
        for (int i = 0; i < nums.size(); i++) {
            string s = to_string(nums[i]);
            string n = "";
            for (char ch : s) {
                n += to_string(mapping[ch - '0']);
            }
            mappedList.push_back(make_tuple(nums[i], stoi(n), i));
        }

        // Step 2: Sort the vector based on the mapped values and original indices for stability
        sort(mappedList.begin(), mappedList.end(), [](const auto& a, const auto& b) {
            if (get<1>(a) != get<1>(b)) {
                return get<1>(a) < get<1>(b);
            } else {
                return get<2>(a) < get<2>(b);
            }
        });

        // Step 3: Create a result vector and fill it with the sorted original nums
        vector<int> sortedNums(nums.size());
        for (int i = 0; i < mappedList.size(); i++) {
            sortedNums[i] = get<0>(mappedList[i]);
        }

        return sortedNums;
    }
};